tests/config/test_logging_config_updater.py

```diff
@@ -0,0 +1,111 @@
+# License: MIT
+# Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+"""Tests for logging config updater."""
+
+import asyncio
+from collections.abc import Mapping
+from typing import Any
+
+import pytest
+from frequenz.channels import Broadcast
+from marshmallow import ValidationError
+from pytest_mock import MockerFixture
+
+from frequenz.sdk.config import LoggerConfig, LoggingConfig, LoggingConfigUpdater
+
+
+def test_logging_config() -> None:
+    """Test if logging config is correctly loaded."""
+    config_raw = {
+        "default": {"level": "DEBUG"},
+        "loggers": {
+            "frequenz.sdk.actor": {"level": "INFO"},
+            "frequenz.sdk.timeseries": {"level": "ERROR"},
+        },
+    }
+    config = LoggingConfig(
+        default=LoggerConfig(level="DEBUG"),
+        loggers={
+            "frequenz.sdk.actor": LoggerConfig(level="INFO"),
+            "frequenz.sdk.timeseries": LoggerConfig(level="ERROR"),
+        },
+    )
+
+    assert LoggingConfig.load(config_raw) == config
+
+    config_raw = {}
+    config = LoggingConfig()
+    assert LoggingConfig.load(config_raw) == config
+
+    config_raw = {"default": {"level": "UNKNOWN"}}
+    with pytest.raises(ValidationError):
+        LoggingConfig.load(config_raw)
+
+    config_raw = {"unknown": {"frequenz.sdk.actor": {"level": "DEBUG"}}}
+    with pytest.raises(ValidationError):
+        LoggingConfig.load(config_raw)
+
+
+async def test_logging_config_updater_actor(
+    mocker: MockerFixture,
+) -> None:
+    """Test if logging is configured and updated correctly."""
+    # Mock logging module in this file. Other logging won't be affected.
+    logging_mock = mocker.patch("frequenz.sdk.config._logging_config_updater.logging")
+
+    config_channel = Broadcast[Mapping[str, Any]](name="config")
+    config_sender = config_channel.new_sender()
+    async with LoggingConfigUpdater(
+        config_recv=config_channel.new_receiver().map(
+            lambda app_config: app_config.get("logging", {})
+        )
+    ) as actor:
+        # Check if default config has been set
+        logging_mock.basicConfig.assert_called_once()
+
+        update_logging_spy = mocker.spy(actor, "_update_logging")
```

Comment:
Nice, I didn't know about `spy`. But with this you are not actually testing the `_update_logging` method itself. I would either change this test or add another test and mock `logging.getLogger` to verify it is being called as expected, or maybe even just instead the real `logging` (basically `l = logging.getLogger(...); asser l.level == expected_level`) to verify the config was actually changed.

Reply:
Done, however things get more complicated. :)
1. python doens't allow to override root level config in unit test (which makes sense because it can mess up a lot).
2. loggings setings for other loggers are not reset after unit test method ends, I had to write teardown method (pytest.fixture) to reset them manually. Please look at test.

Also I found bug - with new config, old loggers were not unset. It is fixed it and tested.

I am not testing root_logging level, but I think it is ok, because then we start testing external library. 
I could check if logging.basicConfig was called with correct arguments, but I am not sure it it valuable check.

Reply:
Yeah, I think for that case you can just mock the `logging.getLogger()` function and just trust that if it is called with the right parameters it will work as expected.